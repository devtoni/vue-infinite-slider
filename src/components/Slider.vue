<template>
  <div class="Slider">
    <b-carousel
      id="carousel-1"
      class="Carousel"
      :interval="3000"
      controls
      fade
      indicators
      background="#ababab"
      @sliding-start="onSlideEnd"
    >
      <b-carousel-slide v-for="(randomPhoto, index) in getRandomPhotos" :key="index">
        <img slot="img" class="d-block w-100" height="720" :src="randomPhoto.url" alt="image slot">
        <p>{{randomPhoto.creditsTo}}</p>
      </b-carousel-slide>
    </b-carousel>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'
import store from '@/store'
import { FETCH_RANDOM_PHOTOS, CLEAR_RANDOM_PHOTOS } from '@/store/actions.type'

export default {
  name: 'Slider',
  data: () => ({
    prevSlide: 0
  }),
  computed: {
    ...mapGetters(['getRandomPhotos'])
  },
  mounted () {
    store.dispatch(FETCH_RANDOM_PHOTOS).then(() => {
      this.isLoaded = true
    })
  },
  methods: {
    async onSlideEnd (slide) {
      if (slide === 0 && this.prevSlide === 3) {
        await this.handleRandomPhotosRenewal()
      } else {
        this.prevSlide = slide - 1 < 0 ? 0 : slide - 1
      }
    },
    async handleRandomPhotosRenewal () {
      await Promise.all([
        store.dispatch(CLEAR_RANDOM_PHOTOS),
        store.dispatch(FETCH_RANDOM_PHOTOS)
      ])
    }
  }
}
</script>

<style lang="scss" scoped>
.Slider {
  width: 80%;
  .Carousel {
    text-shadow: 1px 1px 2px #333;
  }
}
</style>
