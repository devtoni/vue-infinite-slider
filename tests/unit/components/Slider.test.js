import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Slider from '@/components/Slider'
import {
  unsplashServiceResultOnFirstCall,
  unsplashServiceResultOnSecondCall
} from '../../fixtures/unsplashService'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Slider should', () => {
  let wrapper
  let actions
  let state
  let getters
  let store
  beforeEach(() => {
    state = {
      randomPhotos: unsplashServiceResultOnFirstCall
    }
    getters = {
      getRandomPhotos: jest
        .fn()
        .mockImplementationOnce(() => unsplashServiceResultOnFirstCall)
        .mockImplementationOnce(() => unsplashServiceResultOnSecondCall)
    }
    actions = {
      FETCH_RANDOM_PHOTOS: jest.fn()
    }
    store = new Vuex.Store({
      state,
      getters,
      actions
    })
    wrapper = shallowMount(Slider, {
      localVue,
      store
    })
  })
  it('displays the data', () => {
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAll('b-carousel-slide-stub').length).toBe(5)
  })
})
