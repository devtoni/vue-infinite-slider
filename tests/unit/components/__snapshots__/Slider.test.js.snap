// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should displays the data 1`] = `
<div class="Slider">
  <b-carousel-stub id="carousel-1" labelprev="Previous Slide" labelnext="Next Slide" labelgotoslide="Goto Slide" labelindicators="Select a slide to display" interval="3000" indicators="true" controls="true" fade="true" background="#ababab" value="0" class="Carousel">
    <b-carousel-slide-stub imgblankcolor="transparent" contenttag="div" captiontag="h3" texttag="p"><img height="720" src="https://images.unsplash.com/photo-1555582882-3adcea584e86?ixlib=rb-1.2.1&amp;q=80&amp;fm=jpg&amp;crop=entropy&amp;cs=tinysrgb&amp;w=1080&amp;fit=max&amp;ixid=eyJhcHBfaWQiOjY3OTM0fQ" alt="image slot" class="d-block w-100">
      <p>Photo by David Billings</p>
    </b-carousel-slide-stub>
    <b-carousel-slide-stub imgblankcolor="transparent" contenttag="div" captiontag="h3" texttag="p"><img height="720" src="https://images.unsplash.com/photo-1553570525-4985c3f932a6?ixlib=rb-1.2.1&amp;q=80&amp;fm=jpg&amp;crop=entropy&amp;cs=tinysrgb&amp;w=1080&amp;fit=max&amp;ixid=eyJhcHBfaWQiOjY3OTM0fQ" alt="image slot" class="d-block w-100">
      <p>Photo by Mae Mu</p>
    </b-carousel-slide-stub>
    <b-carousel-slide-stub imgblankcolor="transparent" contenttag="div" captiontag="h3" texttag="p"><img height="720" src="https://images.unsplash.com/photo-1554935456-8b221f01dd2b?ixlib=rb-1.2.1&amp;q=80&amp;fm=jpg&amp;crop=entropy&amp;cs=tinysrgb&amp;w=1080&amp;fit=max&amp;ixid=eyJhcHBfaWQiOjY3OTM0fQ" alt="image slot" class="d-block w-100">
      <p>Photo by Roberto Nickson</p>
    </b-carousel-slide-stub>
    <b-carousel-slide-stub imgblankcolor="transparent" contenttag="div" captiontag="h3" texttag="p"><img height="720" src="https://images.unsplash.com/photo-1553978650-0b59ecf2c36f?ixlib=rb-1.2.1&amp;q=80&amp;fm=jpg&amp;crop=entropy&amp;cs=tinysrgb&amp;w=1080&amp;fit=max&amp;ixid=eyJhcHBfaWQiOjY3OTM0fQ" alt="image slot" class="d-block w-100">
      <p>Photo by Janis Karkossa</p>
    </b-carousel-slide-stub>
    <b-carousel-slide-stub imgblankcolor="transparent" contenttag="div" captiontag="h3" texttag="p"><img height="720" src="https://images.unsplash.com/photo-1555157054-4ee7fedaa49c?ixlib=rb-1.2.1&amp;q=80&amp;fm=jpg&amp;crop=entropy&amp;cs=tinysrgb&amp;w=1080&amp;fit=max&amp;ixid=eyJhcHBfaWQiOjY3OTM0fQ" alt="image slot" class="d-block w-100">
      <p>Photo by Brandon VÃ¡zquez</p>
    </b-carousel-slide-stub>
  </b-carousel-stub>
</div>
`;
